### PR TITLE
feat(sdk): add support for IfPresentPlaceholder and ConcatPlaceholder strings

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
Adds support for lossless read and write of IfPresentPlaceholder and ConcatPlaceholder strings.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull 
request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/compiler/pipeline_spec_builder.py [hunk 1]
```diff
@@ -1333,3 +1336,33 @@ def get_inputs(task_group: tasks_group.TasksGroup,
     )
 
     return pipeline_spec
+
+
+def write_pipeline_spec_to_file(pipeline_spec: pipeline_spec_pb2.PipelineSpec,
```
Comment (chensun) [inline sdk/python/kfp/compiler/pipeline_spec_builder.py:1341]: Is this from some other change? Seems irrelevant to this PR?

File: sdk/python/kfp/compiler/pipeline_spec_builder_test.py [hunk 2]
```diff
@@ -199,5 +203,55 @@ def test(self, pipeline_name, is_valid):
...
+
+    def test_incorrect_extension(self):
+        with tempfile.TemporaryDirectory() as tempdir, self.assertRaisesRegex(
+                ValueError, r'should end with "\.yaml"\.'):
+            temp_filepath = os.path.join(tempdir, 'output.txt')
+            pipeline_spec_builder.write_pipeline_spec_to_file(
+                self.pipeline_spec, temp_filepath)
+
+
+if __name__ == '__main__':
```
Comment (chensun) [inline sdk/python/kfp/compiler/pipeline_spec_builder_test.py:253]: dupe

File: sdk/python/kfp/compiler/pipeline_spec_builder.py (hunk 1 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/pipeline_spec_builder.py]: Yes, good catch. This is duplicate code. I suspect the dupe is attributed to a mistake I made when rebasing this PR.

**Review by chensun — COMMENTED** [review]
/lgtm
/approve

Thanks!
